src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py

```diff
@@ -0,0 +1,475 @@
+    CapacityMetrics,
+    PowerBounds,
+    PowerBoundsMetrics,
+    SoC_Capacity,
+    SoC_Capacity_Metrics,
+    SoCMetrics,
+)
+
+_logger = logging.getLogger(__name__)
+
+_MIN_TIMESTAMP = datetime.min.replace(tzinfo=timezone.utc)
+
+# Formula output types class have no common interface
+# Print all possible types here.
+T = TypeVar("T", Sample, SoC_Capacity, PowerBounds)
+
+
+class MetricCalculator(ABC, Generic[T]):
+    """Define how to calculate high level metrics from many components data.
+
+    It specifies:
+        * what components and metrics its needs to calculate the results,
+        * how to calculate the result,
+    """
+
+    def __init__(self, batteries: Set[int]) -> None:
+        """Create class instance
+
+        Args:
+            batteries: From what batteries the data should be aggregated.
+        """
+        self._batteries = batteries
+
+    @classmethod
+    @abstractmethod
+    def name(cls) -> str:
+        """Return name of the formula.
+
+        Returns:
+            Name of the formula
+        """
+
+    @property
+    @abstractmethod
+    def battery_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each battery.
+
+        Returns:
+            Map between battery id and set of required metrics id.
+        """
+
+    @property
+    @abstractmethod
+    def inverter_metrics(self) -> Dict[int, List[ComponentMetricId]]:
+        """Return what metrics are needed for each inverter.
+
+        Returns:
+            Map between inverter id and set of required metrics id.
+        """
+
+    @abstractmethod
+    def calculate(
+        self,
+        metrics_data: Dict[int, ComponentMetricsData],
+        working_batteries: Set[int],
+    ) -> Optional[T]:
+        """Aggregate the metrics_data and calculate high level metric.
+
+        Missing components will be ignored. Formula will be calculated for all
+        working batteries that are in metrics_data.
+
+        Args:
+            metrics_data: Components metrics data, that should be used to calculate the
+                result.
+            working_batteries: working batteries. These batteries will be used
+                to calculate the result. It should be subset of the batteries given in a
+                constructor.
+
+        Returns:
+            High level metric calculated from the given metrics.
+            Return None if there are no component metrics.
+        """
+
+    def _is_missing(
+        self, component_id: int, components_metrics: Dict[int, ComponentMetricsData]
+    ) -> bool:
+        if component_id not in components_metrics:
+            _logger.warning(
+                "%s: Missing metrics from the components %d",
+                self.name,
+                component_id,
+            )
+            return True
+
+        return False
+
+
+# Soc and capacity are different kind of metrics.
+# It makes sens to separate them with `_`.
+class SoC_Capacity_Calculator(
```

Comment:
Why underscore in class names here?

Reply:
I separated it as we discussed on the meeting